src/common/http_client.rs

```diff
@@ -0,0 +1,122 @@
+use std::path::Path;
+use std::{fs, io, result};
+
+use storage::content_manager::errors::StorageError;
+
+use crate::settings::{Settings, TlsConfig};
+
+pub struct HttpClient {
+    tls_config: Option<TlsConfig>,
+    verify_https_client_certificate: bool,
+}
+
+impl HttpClient {
+    pub fn from_settings(settings: &Settings) -> Result<Self> {
+        let tls_config = if settings.service.enable_tls {
+            let Some(tls_config) = settings.tls.clone() else {
+                return Err(Error::TlsConfigUndefined);
+            };
+
+            Some(tls_config)
+        } else {
+            None
+        };
+
+        let verify_https_client_certificate = settings.service.verify_https_client_certificate;
+
+        let http_client = Self {
+            tls_config,
+            verify_https_client_certificate,
+        };
+
+        Ok(http_client)
+    }
+
+    pub fn client(&self) -> Result<reqwest::Client> {
+        match &self.tls_config {
+            Some(tls_config) => https_client(tls_config, self.verify_https_client_certificate),
+            None => Ok(reqwest::Client::new()),
+        }
+    }
+}
+
+fn https_client(
+    tls_config: &TlsConfig,
+    verify_https_client_certificate: bool,
+) -> Result<reqwest::Client> {
+    let mut builder = reqwest::Client::builder()
+        .add_root_certificate(https_client_ca_cert(tls_config.ca_cert.as_ref())?);
+
+    if verify_https_client_certificate {
+        builder = builder.identity(https_client_identity(
+            tls_config.cert.as_ref(),
+            tls_config.key.as_ref(),
+        )?);
+    }
+
+    let client = builder.build()?;
+
+    Ok(client)
+}
+
+fn https_client_ca_cert(ca_cert: &Path) -> Result<reqwest::tls::Certificate> {
+    let ca_cert_pem =
+        fs::read(ca_cert).map_err(|err| Error::failed_to_read(err, "CA certificate", ca_cert))?;
+
+    let ca_cert = reqwest::Certificate::from_pem(&ca_cert_pem)?;
+
+    Ok(ca_cert)
+}
+
+fn https_client_identity(cert: &Path, key: &Path) -> Result<reqwest::tls::Identity> {
+    let mut identity_pem =
+        fs::read(cert).map_err(|err| Error::failed_to_read(err, "certificate", cert))?;
+
+    let mut key_file = fs::File::open(key).map_err(|err| Error::failed_to_read(err, "key", key))?;
+
+    io::copy(&mut key_file, &mut identity_pem)
+        .map_err(|err| Error::failed_to_read(err, "key", key))?;
```

Comment:
Why do we replace the identity bytes (cert) with the key material here?

Reply:
We don't replace, we _append_.

Reply:
`reqwest::tls::Identity` takes a single PEM that should contain both the certificate and the key, and seems like PEM files can be simply concatenated.

Reply:
🤦‍♂️ Thanks for elaborating